includes/class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,336 @@
+<?php
+/**
+ * File containing the class Sensei_Course_Enrolment.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Handles course enrolment logic.
+ */
+class Sensei_Course_Enrolment {
+	const META_PREFIX_ENROLMENT_RESULTS = 'course-enrolment-';
+
+	/**
+	 * Courses instances.
+	 *
+	 * @var static[]
+	 */
+	private static $instances = [];
+
+	/**
+	 * All course enrolment providers.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private static $enrolment_providers;
+
+	/**
+	 * Enrolment providers handling this particular course.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private $course_enrolment_providers;
+
+	/**
+	 * Course enrolment providers version hash.
+	 *
+	 * @var string
+	 */
+	private $course_enrolment_providers_version;
+
+	/**
+	 * Course ID for this enrolment object.
+	 *
+	 * @var int
+	 */
+	private $course_id;
+
+	/**
+	 * Sensei_Course_Enrolment constructor.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 */
+	private function __construct( $course_id ) {
+		$this->course_id = $course_id;
+	}
+
+	/**
+	 * Get instance for a particular course.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 *
+	 * @return self
+	 */
+	public static function get_course_instance( $course_id ) {
+		if ( ! isset( self::$instances[ $course_id ] ) ) {
+			self::$instances[ $course_id ] = new static( $course_id );
+		}
+
+		return self::$instances[ $course_id ];
+	}
+
+	/**
+	 * Gets the course ID for this enrolment object.
+	 *
+	 * @return int
+	 */
+	public function get_course_id() {
+		return $this->course_id;
+	}
+
+	/**
+	 * Check if a user is enroled in a course.
+	 *
+	 * @param int  $user_id     User ID.
+	 * @param bool $check_cache Check and use cached result.
+	 *
+	 * @return bool
+	 */
+	public function is_enroled( $user_id, $check_cache = true ) {
```

Comment:
Oddly enough, the past tense of _enrol_ is _enrolled_.

Reply:
I thought so too but then got mislead by this https://simple.wiktionary.org/wiki/enroled. As a native speaker of this variation I'll trust you, though 😃.

Reply:
643a3766ec430202d23d49c04c77763b965f95cb 

I'll update the WCPC WIPs now.